src/Aspire.Hosting.Analyzers/AppHostAnalyzer.cs

```diff
@@ -0,0 +1,147 @@
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public partial class AppHostAnalyzer : DiagnosticAnalyzer
+{
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => Diagnostics.SupportedDiagnostics;
+
+    public override void Initialize(AnalysisContext context)
+    {
+        context.EnableConcurrentExecution();
+        context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.ReportDiagnostics);
+        context.RegisterCompilationStartAction(AnalyzeCompilationStart);
+    }
+
+    private void AnalyzeCompilationStart(CompilationStartAnalysisContext context)
+    {
+        var compilation = context.Compilation;
+        var wellKnownTypes = WellKnownTypes.GetOrCreate(compilation);
+
+        // We want ConcurrentHashSet here in case RegisterOperationAction runs in parallel.
+        // Since ConcurrentHashSet doesn't exist, use ConcurrentDictionary and ignore the value.
+        var concurrentQueue = new ConcurrentQueue<ConcurrentDictionary<ModelNameOperation, byte>>();
+        context.RegisterOperationBlockStartAction(context =>
+        {
+            // Pool and reuse lists for each block.
+            if (!concurrentQueue.TryDequeue(out var modelNameOperations))
+            {
+                modelNameOperations = new ConcurrentDictionary<ModelNameOperation, byte>();
+            }
+
+            context.RegisterOperationAction(c => DoOperationAnalysis(c, modelNameOperations), OperationKind.Invocation);
+
+            context.RegisterOperationBlockEndAction(c =>
+            {
+                DetectInvalidModelNames(c, modelNameOperations);
+
+                // Return to the pool.
+                modelNameOperations.Clear();
+                concurrentQueue.Enqueue(modelNameOperations);
+            });
+        });
+
+        void DoOperationAnalysis(OperationAnalysisContext context, ConcurrentDictionary<ModelNameOperation, byte> modelNameOperations)
+        {
+            var invocation = (IInvocationOperation)context.Operation;
+            var targetMethod = invocation.TargetMethod;
+
+            if (!IsModelNameInvocation(wellKnownTypes, targetMethod, out var parameterData))
+            {
+                return;
+            }
+
+            if (!TryGetStringToken(invocation, parameterData.Value.ModelNameParameter, out var token))
+            {
+                return;
+            }
+
+            modelNameOperations.TryAdd(ModelNameOperation.Create(invocation, parameterData.Value.ModelType, token), value: default);
+        }
+    }
+
+    private static bool IsModelNameInvocation(WellKnownTypes wellKnownTypes, IMethodSymbol targetMethod, [NotNullWhen(true)] out (IParameterSymbol ModelNameParameter, ModelType ModelType)? parameterData)
+    {
+        // Look for first string parameter annotated with attribute that implements IModelNameParameter
+        ModelType modelType = default;
+        var candidateParameter = targetMethod.Parameters.FirstOrDefault(ps =>
```

Comment:
Is only the first parameter with a model name attribute validated? If all methods only have one at the moment, it's problem just a matter of time until we have a situation with multiple.

Reply:
Right now, yes. I could update it to support multiple.

Reply:
I updated it to support multiple parameters in a single method and multiple attributes on a single parameter.